### PR TITLE
UCP/AM: Limit active message id to uint16_t

### PR DESCRIPTION
**Pull request by amastbaum**
## What
When registering an AM handler, ensure the id fits within the `uint16_t` range.

## Why ?
The AM API accepts the AM handler id as an `unsigned int`, but internally the id is constrained to `uint16_t`. This difference can cause multiple ids (up to 2^16) to map to the same internal id, ultimately referring to the same AM handler (the last one registered).

## How ?
When registering AM handler, ensure `param->id`'s 16 msb are not set. If any of them are non-zero, the function returns an error.

### CODE REVIEW

**Comment by amastbaum** [comment]
@gleon99 it's ready for review (I think I cannot edit the reviewers in the PR).

File: src/ucp/core/ucp_am.c [hunk 1]
```diff
@@ -951,6 +955,11 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nbx,
 
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
 
+    if (id & 0xffff0000) {
+        ret = UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
```
Comment (iyastreb) [inline src/ucp/core/ucp_am.c:959]: I guess we need to log an error with `ucs_error` macro

File: src/ucp/core/ucp_am.c [hunk 2]
```diff
@@ -328,6 +328,10 @@ ucs_status_t ucp_worker_set_am_recv_handler(ucp_worker_h worker,
         return UCS_ERR_INVALID_PARAM;
     }
 
+    if (param->id & 0xffff0000) {
```
Comment (iyastreb) [inline src/ucp/core/ucp_am.c:331]: minor: maybe `if (param->id != (uint16_t)param->id)` to make it more explicit?

File: test/gtest/ucp/test_ucp_am.cc [hunk 3]
```diff
@@ -692,6 +691,148 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+        if (flags != 0) {
+            param.field_mask |= UCP_AM_HANDLER_PARAM_FIELD_FLAGS;
+            param.flags       = flags;
+        }
+
+        return ucp_worker_set_am_recv_handler(e.worker(), &param);
+    }
+
+    void test_am_id_handler(size_t size, size_t header_size = 0ul,
+                            unsigned flags = 0, unsigned data_cb_flags = 0,
```
Comment (iyastreb) [inline test/gtest/ucp/test_ucp_am.cc:733]: I guess we don't really need all default arguments

**Comment by gleon99** [comment]
> amastbaum

@amastbaum plz check your email, added you to maintainers group.

File: src/ucp/core/ucp_am.c (hunk 1 above)
Comment (amastbaum) [reply src/ucp/core/ucp_am.c:959]: I see that using `ucs_error` fails the test automatically (I wanted to specifically check for a failure when the id is too large). @iyastreb  What is preferable? issuing a ucs_error message or testing this case? or is there another solution?

File: src/ucp/core/ucp_am.c (hunk 2 above)
Comment (tvegas1) [reply src/ucp/core/ucp_am.c:331]: `if (param->id > UINT16_MAX)`

File: src/ucp/core/ucp_am.c (hunk 1 above)
Comment (tvegas1) [reply src/ucp/core/ucp_am.c:959]: for test to ignore error, you could use `scoped_log_handler slh(hide_errors_logger);`

Comment (tvegas1) [reply src/ucp/core/ucp_am.c:959]: maybe use UINT16_MAX

File: test/gtest/ucp/test_ucp_am.cc [hunk 4]
```diff
@@ -619,7 +619,6 @@ class test_ucp_am_nbx : public test_ucp_am_base {
         return am_data_rndv_handler(data, length);
     }
 
-
```
Comment (tvegas1) [inline test/gtest/ucp/test_ucp_am.cc:622]: seems unrelated

File: src/ucp/core/ucp_am.c [hunk 5]
```diff
@@ -328,6 +328,12 @@ ucs_status_t ucp_worker_set_am_recv_handler(ucp_worker_h worker,
         return UCS_ERR_INVALID_PARAM;
     }
 
+    if (param->id > UINT16_MAX) {
+        ucs_error("AM handler id is too large (%u). "
```
Comment (iyastreb) [inline src/ucp/core/ucp_am.c:332]: I would rephrase it like:
`AM handler id %u is outside of allowed range [0, %u], id, UINT16_MAX`
But up to you

File: test/gtest/ucp/test_ucp_am.cc [hunk 6]
```diff
@@ -692,6 +692,141 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+        mem_buffer sbuf(size, tx_memtype());
+        sbuf.pattern_fill(SEED);
+        m_hdr.resize(header_size);
+        ucs::fill_random(m_hdr);
+        reset_counters();
+        ucp_mem_h memh = NULL;
+
+        EXPECT_EQ(UCS_OK, set_am_id_overflow_data_handler(receiver(),
+                                        0x1, am_id_overflow_data_cb_1,
+                                        this, 0));
```
Comment (iyastreb) [inline test/gtest/ucp/test_ucp_am.cc:744]: I think you don't need to pass 2 last parameters - `this` can be taken from the `set_am_id_overflow_data_handler`
And `flags` are not used, so no need to have that param at all

File: test/gtest/ucp/test_ucp_am.cc [hunk 7]
```diff
@@ -692,6 +692,141 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+        param.op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE;
+        param.datatype     = sdt_desc.dt();
+
+        if (memh != NULL) {
+            param.op_attr_mask |= UCP_OP_ATTR_FIELD_MEMH;
+            param.memh          = memh;
+        }
+
+        m_send_counter++;
+        ucs_status_ptr_t sptr = ucp_am_send_nbx(sender().ep(), 0x1, m_hdr.data(),
```
Comment (iyastreb) [inline test/gtest/ucp/test_ucp_am.cc:765]: Since we are testing AM IDs, I'm not sure if we need to register and allocate memory.
I would propose to implement the minimal test just for this particular improvement without memory registration/flags/mem_buffer.

File: test/gtest/ucp/test_ucp_am.cc [hunk 8]
```diff
@@ -692,6 +692,111 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+                  !!(rx_param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP));
+        EXPECT_EQ(has_reply_ep, rx_param->reply_ep != NULL);
+
+        if (1 == handler_num) m_recv_counter_cb_1++;
+        else if (2 == handler_num) m_recv_counter_cb_2++;
+
+        return UCS_OK;
+    }
+
+    static ucs_status_t am_id_overflow_data_cb_1(void *arg, const void *header,
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_am.cc:776]: Code repetition.
Maybe use lambda or `std::bind` to eliminate  it?

File: test/gtest/ucp/test_ucp_am.cc [hunk 9]
```diff
@@ -692,6 +692,111 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+protected:
+
+    void reset_counters()
+    {
+        test_ucp_am_nbx::reset_counters();
+        m_recv_counter_cb_1 = 0;
+        m_recv_counter_cb_2 = 0;
+    }
+
+    ucs_status_t set_am_id_overflow_data_handler(entity &e, unsigned am_id,
```
[the remark this answers is not in the document]
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:711]: It's different in that `set_am_data_handler` receives `am_id` as `uint16_t` instead of `unsigned int`, and it also asserts successful registration of data handler while `set_am_id_overflow_data_handler` only returns status, so the opposite case can also be tested when using this function. I can change its name though as there is no overflow here.. or maybe I can change `set_am_data_handler` so it would be more generic.
What do you think?

File: test/gtest/ucp/test_ucp_am.cc [hunk 10]
```diff
@@ -692,6 +692,111 @@ class test_ucp_am_nbx : public test_ucp_am_base {
     ucp_mem_h                       m_rx_memh;
 };
 
+class test_ucp_am_id : public test_ucp_am_nbx {
+public:
+    test_ucp_am_id()
```
[the remark this answers is not in the document]
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:697]: I guess I was trying to mimic the behavior of `test_ucp_am_nbx` and call `reset_counters()` as done there, but on a second thought `reset_counters()` is called at the beginning of `test_am_id_handler()` anyway..

File: test/gtest/ucp/test_ucp_am.cc [hunk 11]
```diff
@@ -692,6 +692,111 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+
+        EXPECT_EQ(UCS_OK, set_am_id_overflow_data_handler(receiver(),
+                                        0x1, am_id_overflow_data_cb_1));
+        EXPECT_NE(UCS_OK, set_am_id_overflow_data_handler(receiver(),
+                                        0xffff0001, am_id_overflow_data_cb_2));
+
+        ucp_request_param_t param;
+        param.op_attr_mask = 0;
+
+        m_send_counter++;
```
[the remark this answers is not in the document]
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:739]: Right.. I just avoided adding a parameter to `update_counter_and_send_am` but for no reason :)

File: test/gtest/ucp/test_ucp_am.cc [hunk 12]
```diff
@@ -692,6 +692,111 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+                           UCP_AM_HANDLER_PARAM_FIELD_CB |
+                           UCP_AM_HANDLER_PARAM_FIELD_ARG;
+        param.id         = am_id;
+        param.cb         = cb;
+        param.arg        = this;
+
+        return ucp_worker_set_am_recv_handler(e.worker(), &param);
+    }
+
+    void test_am_id_handler(size_t size, size_t header_size = 0ul)
```
[the remark this answers is not in the document]
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:727]: no reason actually

File: test/gtest/ucp/test_ucp_am.cc (hunk 8 above)
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:776]: @gleon99 I was trying to use either lambda or `std::bind` to define a function of type `ucp_am_recv_callback_t` but it didn't work and became too messy.. I am pushing the code without this change for now, but would like to talk about it, maybe I have some gaps in this matter

File: test/gtest/ucp/test_ucp_am.cc [hunk 13]
```diff
@@ -692,6 +693,69 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+        test_ucp_am_nbx::reset_counters();
+        m_recv_counter_cb_1 = 0;
+        m_recv_counter_cb_2 = 0;
+    }
+
+    void test_am_id_handler()
+    {
+        reset_counters();
+
+        EXPECT_UCS_OK(test_ucp_am_nbx::set_am_data_handler(
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_am.cc:709]: `test_ucp_am_nbx::` really needed?

File: test/gtest/ucp/test_ucp_am.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (gleon99) [reply test/gtest/ucp/test_ucp_am.cc]: I see.
Than maybe have `set_am_data_handler` in the base class, which would call `set_am_data_handler_internal` - which will do the job and return rc.
`set_am_data_handler` will `ASSERT_...` so that no change in existing code is needed.

Your new func `set_am_..something` in the subclass will call `set_am_data_handler_internal` as well.

[the remark this answers is not in the document]
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc]: Oh, I missed this comment yesterday. I can revert it

File: test/gtest/ucp/test_ucp_am.cc (hunk 8 above)
Comment (gleon99) [reply test/gtest/ucp/test_ucp_am.cc:732]: Per our f2f discussion, lambda or `std::bind` won't work here because of C/C++ types interoperability limitation.

File: src/ucp/core/ucp_am.c [hunk 14]
```diff
@@ -951,6 +957,13 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nbx,
 
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
 
+    if (id > UINT16_MAX) {
```
Comment (brminich) [inline src/ucp/core/ucp_am.c:960]: maybe 
```suggestion
    if (ENABLE_PARAMS_CHECK && ucs_unlikely(id > UINT16_MAX)) {
```
to avoid extra overhead on fast path, like it is done in other funcs in this file (e.g. `ucp_am_recv_data_nbx`)

File: test/gtest/ucp/test_ucp_am.cc [hunk 15]
```diff
@@ -692,6 +700,69 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+    }
+
+    void test_am_id_handler()
+    {
+        reset_counters();
+
+        EXPECT_UCS_OK(set_am_data_handler_internal(
+                                        receiver(), 0x1,
+                                        am_id_overflow_data_cb_1, this));
+        EXPECT_NE(UCS_OK, set_am_data_handler_internal(
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:719]:
```suggestion
        EXPECT_EQ(UCS_ERR_INVALID_PARAM, set_am_data_handler_internal(
```

File: src/ucp/core/ucp_am.c (hunk 14 above)
Comment (amastbaum) [reply src/ucp/core/ucp_am.c:960]: @brminich When is `ENABLE_PARAMS_CHECK` set to 0?

Comment (brminich) [reply src/ucp/core/ucp_am.c]: it is a configure option which is set to `0` for release builds
```
# ./configure --help | grep params
  --disable-params-check  Disable checking user parameters passed to API,
# cat ./contrib/configure-release | grep params
        --disable-params-check \
```

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_am.c (hunk 14 above)
Comment (amastbaum) [reply src/ucp/core/ucp_am.c]: But I think we want to check this parameter in release as well. Don't we?

Comment (amastbaum) [reply src/ucp/core/ucp_am.c]: I see. Thanks

File: src/ucp/core/ucp_am.c [hunk 16]
```diff
@@ -328,6 +328,12 @@ ucs_status_t ucp_worker_set_am_recv_handler(ucp_worker_h worker,
         return UCS_ERR_INVALID_PARAM;
     }
 
+    if (param->id > UINT16_MAX) {
+        ucs_error("Invalid AM id %u, must be in range [0, %u]",
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:332]: small case

File: src/ucp/core/ucp_am.c [hunk 17]
```diff
@@ -951,6 +957,13 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nbx,
 
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
 
+    if (ENABLE_PARAMS_CHECK && ucs_unlikely(id > UINT16_MAX)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
+        ucs_error("Invalid AM id %u, must be in range [0, %u]",
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:962]: 1. small case
2. try to unite common code with ucp_worker_set_am_recv_handler(), a helper function like ucp_am_check_id(unsigned am_id) that will print the error message and return failure status if am id is bad

File: test/gtest/ucp/test_ucp_am.cc [hunk 18]
```diff
@@ -692,6 +700,69 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+        test_ucp_am_nbx::reset_counters();
+        m_recv_counter_cb_1 = 0;
+        m_recv_counter_cb_2 = 0;
+    }
+
+    void test_am_id_handler()
+    {
+        reset_counters();
+
+        EXPECT_UCS_OK(set_am_data_handler_internal(
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:716]: can were just use set_am_data_handler?

File: test/gtest/ucp/test_ucp_am.cc [hunk 19]
```diff
@@ -692,6 +700,69 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+                                              NULL, 0ul, NULL, 0, 0x1, &param);
+        wait_receives();
+        ASSERT_UCS_OK(request_wait(sptr));
+        EXPECT_EQ(m_recv_counter_cb_1, 1);
+        EXPECT_EQ(m_recv_counter_cb_2, 0);
+
+        sptr = update_counter_and_send_am(NULL, 0ul, NULL, 0, 0xffff0001, &param);
+        EXPECT_EQ(UCS_PTR_STATUS(sptr), UCS_ERR_INVALID_PARAM);
+        EXPECT_EQ(m_recv_counter_cb_1, 1);
+        EXPECT_EQ(m_recv_counter_cb_2, 0);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:736]: I don't think we need to test the "good" flow, there are quite a lot of ucp ap tests already, in the new tests we can just check the erroneous cases of passing invalid am id

File: test/gtest/ucp/test_ucp_am.cc (hunk 18 above)
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:716]: set_am_data_handler only receives uint16_t as a handler id (this is the original implementation, and it was suggested to keep it that way although now I am not sure why), and I wanted to reproduce a scenario where the user passes 2 different ids that are eventually translated to the same handler function, and for that I need to use a function that receives unsigned int.

Comment (yosefe) [reply test/gtest/ucp/test_ucp_am.cc:716]: I think we need to change set_am_data_handler to accept unsigned am_im

File: src/ucp/core/ucp_am.c (hunk 17 above)
Comment (amastbaum) [reply src/ucp/core/ucp_am.c:962]: 2. so we prefer to use ENABLE_PARAMS_CHECK condition on both cases? (fast **and slow** path)?

File: test/gtest/ucp/test_ucp_am.cc (hunk 18 above)
Comment (amastbaum) [reply test/gtest/ucp/test_ucp_am.cc:716]: Another difference I forgot to mention here is that set_am_data_handler_internal doesn't check/expect success while set_am_data_handler does. All the other functions that call set_am_data_handler assume it succeeds except for this function (test_am_id_handler) so this is another reason for dividing it into 2 functions.

File: src/ucp/core/ucp_am.c (hunk 17 above)
Comment (amastbaum) [reply src/ucp/core/ucp_am.c:962]: @yosefe 
2. after adding a helper function
```
static ucs_status_t ucp_am_check_id(unsigned am_id)
{
    if (ENABLE_PARAMS_CHECK && ucs_unlikely(am_id > UINT16_MAX)) {
        printf("invalid AM id %u, must be in range [0, %u]",
                  am_id, UINT16_MAX);
        return UCS_ERR_INVALID_PARAM;
    }

    return UCS_OK;
}
```
I get a very strange warning that fails gtest and it seems to be unrelated (at least not directly related).
This is the output:

```
$ GTEST_FILTER=*rc/*test_ucp_am_id.am_id_overflow* ./test/gtest/gtest

[     INFO ] ugni is not available
[     INFO ] ugni,cuda_copy,rocm_copy is not available
[     INFO ] Using random seed of 10211
Note: Google Test filter = *rc/*test_ucp_am_id.am_id_overflow*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from rc/test_ucp_am_id
[ RUN      ] rc/test_ucp_am_id.am_id_overflow/0 <rc_v>
invalid AM id 4294901761, must be in range [0, 65535]
invalid AM id 4294901761, must be in range [0, 65535]
[1719478243.466888] [rock28:19096:0]        spinlock.c:29   UCX  WARN  ucs_recursive_spinlock_destroy() failed: busy
common/test.cc:366: Failure
Failed
Got 0 errors and 1 warnings during the test
[     INFO ] < type/spinlock.c:29 ucs_recursive_spinlock_destroy() failed: busy >
[  FAILED  ] rc/test_ucp_am_id.am_id_overflow/0, where GetParam() = rc_v (347 ms)
[ RUN      ] rc/test_ucp_am_id.am_id_overflow/1 <rc_v/proto_v1>
invalid AM id 4294901761, must be in range [0, 65535]
invalid AM id 4294901761, must be in range [0, 65535]
[1719478243.741695] [rock28:19096:0]        spinlock.c:29   UCX  WARN  ucs_recursive_spinlock_destroy() failed: busy
common/test.cc:366: Failure
Failed
Got 0 errors and 1 warnings during the test
[     INFO ] < type/spinlock.c:29 ucs_recursive_spinlock_destroy() failed: busy >
[  FAILED  ] rc/test_ucp_am_id.am_id_overflow/1, where GetParam() = rc_v/proto_v1 (274 ms)
[----------] 2 tests from rc/test_ucp_am_id (621 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (623 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] rc/test_ucp_am_id.am_id_overflow/0, where GetParam() = rc_v
[  FAILED  ] rc/test_ucp_am_id.am_id_overflow/1, where GetParam() = rc_v/proto_v1

 2 FAILED TESTS
```

Have you seen something like this?

Unless it's a silly mistake and I missed something, please let me know if you want me to keep searching for the cause in the scope of this PR.

File: test/gtest/ucp/test_ucp_am.cc [hunk 20]
```diff
@@ -692,6 +700,44 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+    {
+        test_ucp_am_nbx::reset_counters();
+        m_recv_counter_cb = 0;
+    }
+
+    void test_am_id_handler()
+    {
+        reset_counters();
+
+        EXPECT_EQ(UCS_ERR_INVALID_PARAM, set_am_data_handler_internal(
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:715]: ucs_status_t status = set_am_data_handler_internal(...)
EXPECT_EQ(UCS_ERR_INVALID_PARAM, status)

File: test/gtest/ucp/test_ucp_am.cc [hunk 21]
```diff
@@ -692,6 +700,44 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+                                         receiver(), 0xffff0001,
+                                         am_id_overflow_data_cb, this));
+
+        ucp_request_param_t param;
+        param.op_attr_mask = 0;
+
+        ucs_status_ptr_t sptr = update_counter_and_send_am(NULL, 0ul, NULL, 0, 0xffff0001, &param);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:722]: line seems too long - pls check clang-format output

File: test/gtest/ucp/test_ucp_am.cc [hunk 22]
```diff
@@ -692,6 +700,44 @@ class test_ucp_am_nbx : public test_ucp_am_base {
...
+        EXPECT_EQ(UCS_PTR_STATUS(sptr), UCS_ERR_INVALID_PARAM);
+        EXPECT_EQ(m_recv_counter_cb, 0);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:724]: EXPECT_EQ(0, m_recv_counter_cb);

**Review by yosefe — APPROVED** [review]
pls squash
